presto-main/src/main/java/io/prestosql/sql/planner/NullityEvaluator.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import io.prestosql.sql.tree.ComparisonExpression;
+import io.prestosql.sql.tree.DefaultExpressionTraversalVisitor;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.Literal;
+import io.prestosql.sql.tree.LogicalBinaryExpression;
+import io.prestosql.sql.tree.Node;
+import io.prestosql.sql.tree.NullLiteral;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Determines whether a given Expression can return null value when evaluated
+ */
+public final class NullityEvaluator
```

Comment:
There is already `NullabilityAnalyzer`. Maybe improve that instead? It has `mayReturnNullOnNonNullInput` which should be more useful.